### DISCUSSION TITLE
Molecular surface theme

### DISCUSSION THREAD

**Discussion by luna215** [molstar/molstar#595] [Q&A]
Hi,

I'm trying to color the molecular-surface type with the atoms charge value. 

I have some representations defined like this: 
```typescript
// ..
const representations = {
            ligand: builder.buildRepresentation(update, components.ligand, { type: 'ball-and-stick', typeParams: { ...typeParams, sizeFactor: 0.1 }, color: 'element-symbol', colorParams: { carbonColor: { name: 'element-symbol', params: {} } } }, { tag: 'ligand' }),
            ballAndStick: builder.buildRepresentation(update, components.surroundings, { type: 'molecular-surface', typeParams: { ...typeParams, alpha: 0.44, material: CustomMaterial, sizeFactor: 0.1, sizeAspectRatio: 0.5 }, color: 'hydrophobicity', colorParams: { carbonColor: { name: 'element-symbol', params: {} } } }, { tag: 'surroundings' }),
        };
//.. 
```
and I replaced the color function in the `hydrophobicity.ts` file with the following: 

```typescript
function color(location: Location): Color {
        if (StructureElement.Location.is(location)) {
            const q = getPartialCharge(location.unit, location.element);
            return q !== undefined ? scale.color(q) : DefaultPartialChargeColor;
        } else if (Bond.isLocation(location)) {
            const q = getPartialCharge(location.aUnit, location.aUnit.elements[location.aIndex]);
            return q !== undefined ? scale.color(q) : DefaultPartialChargeColor;
        }
        return DefaultPartialChargeColor;
    }
```
but I keep getting the charge as `undefined`. Does this mean that `molecular-surface` needs to be modified in order to work with atoms? Any guidance/help here is greatly appreciated!

**Comment by luna215** [comment]
Update:
I created a custom theme provider and here is my color function:

```typescript
color: function (location) {
            console.log('location', location);
            if (StructureElement.Location.is(location)) {
                var q = getPartialCharge(location.unit, location.element);
                console.log('structure element', q);
                return q !== undefined ? scale.color(q) : DefaultPartialChargeColor;
            }
            else if (Bond.isLocation(location)) {
                var q = getPartialCharge(location.aUnit, location.aUnit.elements[location.aIndex]);
                return q !== undefined ? scale.color(q) : DefaultPartialChargeColor;
            }
            return DefaultPartialChargeColor;
        },
```

However, my location is being returned as a location of `kind: 'position-location'`. I'm trying to figure out how to convert this to a location of type StructureElement.location 🤔

**Reply by dsehnal** [reply]
You will need to figure out the charge from the 3D position. For example, by creating a 3D grid. It is also possible to create the grid and pass it to color the surface directly in the shader.

**Reply by dsehnal** [reply]
You can also use the `palette` to achieve the coloring gradient that is useful for the charges (see `src\examples\basic-wrapper\custom-theme.ts`)

**Reply by luna215** [reply]
@dsehnal always appreciate your replies! But I'm a bit confused here, do you mean that this functionality, will return the charge from the 3D position:

```typescript
color: location => {
    if (!isPositionLocation(location)) return ColorNames.black;
    const dist = Vec3.squaredDistance(location.position, center);
    const t = Math.min(dist / radiusSq, 1);
    return ((t * scale) | 0) as Color;
}
```

Is there an example of creating a 3D grid?

**Reply by dsehnal** [reply]
No, that function will not return the charges. You would still have to compute the grid or sample the charge based on the distance yourself.

But now that I think about it, element symbol coloring does transfer atom-based coloring to surfaces, so you should be able to use that as the basis for you theme.

----

Unless of course you are trying to color a custom surface, then you need the sampling...

**Comment by arose** [comment]
Make sure the granularity of your theme is group not vertex when you want to use element location.

**Comment by luna215** [comment]
Thanks! Yup, finally got the atoms charge to show up! My two mistakes:
- granularity was set to `vertex` instead  of `group`
- I was applying the `color: electrostatic` to the surroundings representation (I'm guessing this is the reason atoms would not show) rather than the ligand representation.
